File: backend/schemas/holding.py
from datetime import datetime
from decimal import Decimal
from typing import Optional, List

from pydantic import BaseModel, Field, ConfigDict


class HoldingCreate(BaseModel):
    asset_id: int
    quantity: Decimal = Field(..., ge=0)
    avg_cost: Optional[Decimal] = Field(None, ge=0)


class HoldingUpdate(BaseModel):
    quantity: Optional[Decimal] = Field(None, ge=0)
    avg_cost: Optional[Decimal] = Field(None, ge=0)


class HoldingRead(BaseModel):
    id: int
    portfolio_id: int
    asset_id: int
    quantity: Decimal
    avg_cost: Optional[Decimal]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class HoldingDetailRead(HoldingRead):
    transactions: List["TransactionRead"] = []

File: frontend/src/types/holding.ts
export type HoldingRead = {
  id: number;
  portfolio_id: number;
  asset_id: number;
  quantity: string | number;
  avg_cost: string | number | null;
  created_at: string;
  updated_at: string;
};

export type HoldingCreateBody = {
  portfolio_id: number;
  asset_id: number;
  quantity: string;
  avg_cost?: string | null;
};

export type HoldingUpdateBody = {
  quantity?: string;
  avg_cost?: string | null;
};
